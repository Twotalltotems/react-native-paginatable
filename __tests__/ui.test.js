import React from 'react'
import { combineReducers, applyMiddleware, compose, createStore } from 'redux'
import ReduxThunk from 'redux-thunk'
import 'react-native'
import { Text } from 'react-native'

import PaginatableList from '../index'
import PaginationStateManager from '../PaginationStateManager'

import TestRenderer from 'react-test-renderer' // comes after react-native

const USER_ITEMS = [{ key: 'k1' }, { key: 'k2' }, { key: 'k3' }]
const STATE_NAME = 'users'
const URL = 'http://test.example.com/users'

describe('PaginatableList', () => {
	const paginationStateManager = new PaginationStateManager(STATE_NAME, URL)
	const getCombinedReducers = () => {
		return combineReducers({
			[STATE_NAME]: paginationStateManager.reducer()
		})
	}

	describe('.defaultProps', () => {
		describe('with no items', () => {
			it('renders empty state properly', () => {
				const reduxStore = createStore(
					getCombinedReducers(),
					{},
					compose(applyMiddleware(ReduxThunk))
				)
				const tree = TestRenderer.create(
					<PaginatableList
						store={reduxStore}
						customizedPaginationStateManager={
							paginationStateManager
						}
					/>
				).toJSON()

				expect(tree).toMatchSnapshot()
			})
		})

		describe('with a few items', () => {
			it('renders list properly', () => {
				const reduxStore = createStore(
					getCombinedReducers(),
					{ [STATE_NAME]: { items: USER_ITEMS } },
					compose(applyMiddleware(ReduxThunk))
				)
				const tree = TestRenderer.create(
					<PaginatableList
						store={reduxStore}
						customizedPaginationStateManager={
							paginationStateManager
						}
					/>
				).toJSON()

				expect(tree).toMatchSnapshot()
			})
		})
	})

	describe('#onRenderItem', () => {
		it('renders each item using onRenderItem', () => {
			const reduxStore = createStore(
				getCombinedReducers(),
				{ [STATE_NAME]: { items: USER_ITEMS } },
				compose(applyMiddleware(ReduxThunk))
			)
			const onRenderItem = ({ item }) => {
				return <Text>Custom onRenderItem: {item.key}</Text>
			}

			const tree = TestRenderer.create(
				<PaginatableList
					onRenderItem={onRenderItem}
					store={reduxStore}
					customizedPaginationStateManager={paginationStateManager}
				/>
			).toJSON()

			expect(tree).toMatchSnapshot()
		})
	})

	describe('#onRenderEmptyStatus', () => {
		it('renders empty state using onRenderEmptyStatus', () => {
			const reduxStore = createStore(
				getCombinedReducers(),
				{},
				compose(applyMiddleware(ReduxThunk))
			)
			const onRenderEmptyStatus = () => (
				<Text>Custom empty component</Text>
			)

			const tree = TestRenderer.create(
				<PaginatableList
					onRenderEmptyStatus={onRenderEmptyStatus}
					store={reduxStore}
					customizedPaginationStateManager={paginationStateManager}
				/>
			).toJSON()

			expect(tree).toMatchSnapshot()
		})
	})
})
