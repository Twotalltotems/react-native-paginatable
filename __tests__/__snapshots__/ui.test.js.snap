// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatableList #onRenderEmptyStatus renders empty state using onRenderEmptyStatus 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={true}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "width": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <Text>
      Custom empty component
    </Text>
  </View>
</RCTScrollView>
`;

exports[`PaginatableList #onRenderItem renders each item using onRenderItem 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  data={
    Array [
      Object {
        "key": "k1",
      },
      Object {
        "key": "k2",
      },
      Object {
        "key": "k3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={true}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "width": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        Custom onRenderItem: 
        k1
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        Custom onRenderItem: 
        k2
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        Custom onRenderItem: 
        k3
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PaginatableList .defaultProps with a few items renders list properly 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  data={
    Array [
      Object {
        "key": "k1",
      },
      Object {
        "key": "k2",
      },
      Object {
        "key": "k3",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={true}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "width": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "borderBottomColor": "#c0c1c4",
            "borderBottomWidth": 1,
            "height": 120,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9f9fa0",
            }
          }
        >
          Item 0
        </Text>
        <Text
          style={
            Object {
              "color": "#9f9fa0",
              "fontSize": 12,
            }
          }
        >
          Use 'onRenderItem' props to overwrite the default cell.
        </Text>
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../assets/TTTLogo_white.png",
            }
          }
          style={
            Object {
              "alignSelf": "flex-end",
              "height": 40,
              "marginTop": 10,
              "tintColor": "#c0c1c4",
              "width": 80,
            }
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "borderBottomColor": "#c0c1c4",
            "borderBottomWidth": 1,
            "height": 120,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9f9fa0",
            }
          }
        >
          Item 1
        </Text>
        <Text
          style={
            Object {
              "color": "#9f9fa0",
              "fontSize": 12,
            }
          }
        >
          Use 'onRenderItem' props to overwrite the default cell.
        </Text>
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../assets/TTTLogo_white.png",
            }
          }
          style={
            Object {
              "alignSelf": "flex-end",
              "height": 40,
              "marginTop": 10,
              "tintColor": "#c0c1c4",
              "width": 80,
            }
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "borderBottomColor": "#c0c1c4",
            "borderBottomWidth": 1,
            "height": 120,
            "padding": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9f9fa0",
            }
          }
        >
          Item 2
        </Text>
        <Text
          style={
            Object {
              "color": "#9f9fa0",
              "fontSize": 12,
            }
          }
        >
          Use 'onRenderItem' props to overwrite the default cell.
        </Text>
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../assets/TTTLogo_white.png",
            }
          }
          style={
            Object {
              "alignSelf": "flex-end",
              "height": 40,
              "marginTop": 10,
              "tintColor": "#c0c1c4",
              "width": 80,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PaginatableList .defaultProps with no items renders empty state properly 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.5}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={true}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "width": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../assets/TTTLogo_white.png",
          }
        }
        style={
          Object {
            "margin": 10,
            "maxHeight": 100,
            "tintColor": "#c0c1c4",
            "width": "50%",
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#9f9fa0",
          }
        }
      >
        There is no items in the list.
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
